myhpom/tests/test_cloudfactory_tasks.py

```diff
@@ -94,10 +97,268 @@ def test_connection_timeout(self, reqmock):
+            self.assertEqual(cf_run.status, CloudFactoryDocumentRun.STATUS_NOTFOUND)
+            cf_run.delete()
+
+    def test_other_response_status_code(self, reqmock):
+        initial_status = CloudFactoryDocumentRun.STATUS_PROCESSING
+        for status_code in [201, 202, 302, 400, 422, 500]:  # some common response statuses
+            cf_run = CloudFactoryDocumentRun.objects.create(
+                status=initial_status, run_id="SOME_RUN"
+            )
+            reqmock.get(
+                settings.CLOUDFACTORY_API_URL + '/runs/' + cf_run.run_id,
+                status_code=status_code,
+                text='{}',
+            )
+            with self.assertRaises(ValueError):
+                self.task(cf_run.id)
+            cf_run.refresh_from_db()
+            self.assertTrue(reqmock.called)
+            self.assertEqual(cf_run.status, initial_status)
+            cf_run.delete()
+
+    def test_no_change(self, reqmock):
+        status_choices = [
+            status
+            for status in [status[0] for status in CloudFactoryDocumentRun.STATUS_CHOICES]
+            if status
+            not in [
+                CloudFactoryDocumentRun.STATUS_PROCESSING,
+                CloudFactoryDocumentRun.STATUS_TIMEOUT,
+                CloudFactoryDocumentRun.STATUS_ABORTED,
+            ]
+        ]
+        for initial_status in status_choices:
+            cf_run = CloudFactoryDocumentRun.objects.create(
+                status=initial_status, run_id="SOME_RUN"
+            )
+            reqmock.get(
+                settings.CLOUDFACTORY_API_URL + '/runs/' + cf_run.run_id,
+                status_code=200,  # doesn't matter, not called
+                text='{}',
+            )
+            self.task(cf_run.id)
+            cf_run.refresh_from_db()
+            self.assertFalse(reqmock.called)
+            self.assertEqual(cf_run.status, initial_status)
+            cf_run.delete()
+
+
+@requests_mock.Mocker()
+@override_settings(CLOUDFACTORY_API_URL='https://TEST.NIL')
+class CloudFactoryAbortDocumentRunTestCase(TestCase):
+    """
+    * if run.status in [STATUS_PROCESSING, STATUS_TIMEOUT], abort the process.
+        * 202 => STATUS_ABORTED
+        * 404 => STATUS_NOTFOUND
+        * 405 => STATUS_PROCESSED
+        * otherwise => ValueError
+    * otherwise, nothing is done.
+    """
+
+    def setUp(self):
+        self.task = CloudFactoryAbortDocumentRun
+
+    def test_202_successful(self, reqmock):
+        for status in [
+            CloudFactoryDocumentRun.STATUS_PROCESSING,
+            CloudFactoryDocumentRun.STATUS_TIMEOUT,
+        ]:
+            cf_run = CloudFactoryDocumentRun.objects.create(status=status, run_id="SOME_RUN")
+            reqmock.post(
+                settings.CLOUDFACTORY_API_URL + '/runs/' + cf_run.run_id + '/abort',
+                status_code=202,
+                text='{}',
+            )
+            self.task(cf_run.id)
+            self.assertTrue(reqmock.called)
```

Comment:
The `reqmock.called` here doesn't assert anything useful inside the loop. I guess you could user `call_count`? A true mock would let you reset the mock...but I don't see anything in the requests-mock library that lets you do that :/

Reply:
hmmm, good point.